src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -74,13 +117,388 @@ public class ConsensusManager
+            {
+                partialValidationRequired = this.chainedHeaderTree.BlockDataDownloaded(chainedHeaderBlock.ChainedHeader, chainedHeaderBlock.Block);
+            }
+
+            if (partialValidationRequired)
+            {
+                this.blockValidator.StartPartialValidation(chainedHeaderBlock, this.OnPartialValidationCompletedCallback);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        private void OnPartialValidationCompletedCallback(ChainedHeaderBlock chainedHeaderBlock, bool success)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:{3})", nameof(chainedHeaderBlock), chainedHeaderBlock, nameof(success), success);
+
+            if (success)
+            {
+                this.OnPartialValidationSucceeded(chainedHeaderBlock);
+            }
+            else
+            {
+                List<int> peersToBan;
+
+                lock (this.treeLock)
+                {
+                    peersToBan = this.chainedHeaderTree.PartialOrFullValidationFailed(chainedHeaderBlock.ChainedHeader);
+                }
+
+                foreach (int peerId in peersToBan)
+                {
+                    // TODO: ban and disconnect those peers
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        private void OnPartialValidationSucceeded(ChainedHeaderBlock chainedHeaderBlock)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeaderBlock), chainedHeaderBlock);
+
+            bool reorged = false;
+            List<ChainedHeader> nextHeadersToValidate = null;
+
+            lock (this.reorgLock)
+            {
+                bool reorgRequired = false;
+
+                lock (this.treeLock)
+                {
+                    nextHeadersToValidate = this.chainedHeaderTree.PartialValidationSucceeded(chainedHeaderBlock.ChainedHeader, out reorgRequired);
+                }
+
+                if (reorgRequired)
+                {
+                    reorged = this.FullyValidateAndReorgLocked();
+                }
+            }
+
+            if (reorged)
+            {
+                List<ConnectNewHeadersResult> blocksToDownload = new List<ConnectNewHeadersResult>();
+
+                var peers = this.connectionManager.ConnectedPeers;
+
+                foreach (INetworkPeer peer in peers)
+                {
+                    List<ChainedHeader> headersToConnect =  peer.Behavior<ChainHeadersBehavior>().ConsensusAdvanced(this.Tip);
+
+                    if (headersToConnect != null)
+                    {
+                        List<BlockHeader> headers = headersToConnect.Select(ch => ch.Header).ToList();
+
+                        lock (this.treeLock)
+                        {
+                            ConnectNewHeadersResult connectNewHeaders = this.chainedHeaderTree.ConnectNewHeaders(peer.Connection.Id, headers);
+                            blocksToDownload.Add(connectNewHeaders);
```

Comment:
I think here we want to add to the list only if download to is not null

Reply:
Well spotted